src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -653,8 +653,19 @@ void GcsActorManager::HandleGetNamedActorInfo(
       iter->second->GetState() == rpc::ActorTableData::DEAD) {
     // The named actor was not found or the actor is already removed.
     std::stringstream stream;
-    stream << "Actor with name '" << name << "' was not found.";
-    RAY_LOG(DEBUG) << stream.str();
+    if (actor_id.IsNil() || iter == registered_actors_.end()) {
+      stream << "Actor with name '" << name << "' was not found.";
+      RAY_LOG(DEBUG) << stream.str();
+    } else {
+      RAY_CHECK(iter->second->GetState() == rpc::ActorTableData::DEAD);
+      stream << "Actor with name '" << name << "' was found, but it is DEAD.";
+      RAY_LOG(DEBUG) << stream.str();
+      // Update `registered_actors_` and `named_actors_` immediately to avoid race
+      // conditions. Without this, GCS might tell the core worker that the actor is not
+      // found when the core worker tries to `get_actor`, but still think the actor exists
+      // when the core worker tries to create a new actor with the same name.
+      DestroyActor(actor_id, GenActorRefDeletedCause(GetActor(actor_id)));
```

Comment:
After #49480 an actor has three states:

1. ALIVE
2. DEAD and non-restartable: not in `registered_actors_` and `named_actors_`
3. DEAD and restartable: in registered_actors_ and named_actors_

The semantic is clear for get_if_exist if the actor is in state 1 or 2. What we need to figure out is the behavior when the actor is in state 3.

Reply:
One option is to bring it back to state 1.